ipld/plugin/nmt.go

```diff
@@ -0,0 +1,406 @@
+
+	// NmtCodecName is the name used during registry of the NmtCodec codec
+	NmtCodecName = "nmt-node"
+
+	// Sha256Namespace8Flagged is the multihash code used to hash blocks
+	// that contain an NMT node (inner and leaf nodes).
+	Sha256Namespace8Flagged = 0x7701
+
+	// DagParserFormatName can be used when putting into the IPLD Dag
+	DagParserFormatName = "extended-square-row-or-col"
+
+	// Repeated here to avoid a dependency to the wrapping repo as this makes
+	// it hard to compile and use the plugin against a local ipfs version.
+	// TODO: plugins have config options; make this configurable instead
+	namespaceSize = 8
+	shareSize     = 256
+	// nmtHashSize is the size of a digest created by an NMT in bytes.
+	nmtHashSize = 2*namespaceSize + sha256.Size
+)
+
+func init() {
+	mustRegisterNamespacedCodec(
+		Sha256Namespace8Flagged,
+		"sha2-256-namespace8-flagged",
+		nmtHashSize,
+		func() hash.Hash {
+			return &namespaceHasher{
+				nmt.NewNmtHasher(sha256.New(), nmt.DefaultNamespaceIDLen, true),
+			}
+		},
+	)
+	// this should already happen when the plugin is injected but it doesn't for some CI tests
+	ipld.DefaultBlockDecoder.Register(NmtCodec, NmtNodeParser)
+	// register the codecs in the global maps
+	cid.Codecs[NmtCodecName] = NmtCodec
+	cid.CodecToStr[NmtCodec] = NmtCodecName
+}
+
+func mustRegisterNamespacedCodec(
+	codec uint64,
+	name string,
+	defaultLength int,
+	hashFunc func() hash.Hash,
+) {
+	if _, ok := mh.Codes[codec]; !ok {
+		// make sure that the Codec wasn't registered from somewhere different than this plugin already:
+		if _, found := mh.Codes[codec]; found {
+			panic(fmt.Sprintf("Codec 0x%X is already present: %v", codec, mh.Codes[codec]))
+		}
+		// add to mh.Codes map first, otherwise mh.RegisterHashFunc would err:
+		mh.Codes[codec] = name
+		mh.Names[name] = codec
+		mh.DefaultLengths[codec] = defaultLength
+
+		mhcore.Register(codec, hashFunc)
+	}
+}
+
+type namespaceHasher struct {
+	hash.Hash
+}
+
+func (n namespaceHasher) Sum(data []byte) []byte {
+	isLeafData := data[0] == nmt.LeafPrefix
+	if isLeafData {
+		return nmt.Sha256Namespace8FlaggedLeaf(data[1:])
+	}
+	return nmt.Sha256Namespace8FlaggedInner(data[1:])
+}
+
+// DataSquareRowOrColumnRawInputParser reads the raw shares and extract the IPLD nodes from the NMT tree.
+// Note, to parse without any error the input has to be of the form:
+//
+// <share_0>| ... |<share_numOfShares - 1>
+//
+// Note while this coredag.DagParser is implemented here so this plugin can be used from
+// the commandline, the ipld Nodes will rather be created together with the NMT
+// root instead of re-computing it here.
+func DataSquareRowOrColumnRawInputParser(r io.Reader, _mhType uint64, _mhLen int) ([]ipld.Node, error) {
+	br := bufio.NewReader(r)
+	collector := newNodeCollector()
+
+	n := nmt.New(
+		sha256.New(),
+		nmt.NamespaceIDSize(namespaceSize),
+		nmt.NodeVisitor(collector.visit),
+	)
+
+	for {
+		namespacedLeaf := make([]byte, shareSize+namespaceSize)
```

Comment:
is this right? I thought we changed it so that the namespace was embedded into the share, so that the total size is 256 bytes. cc @evan-forbes

Reply:
It should be shareSize + namespaceSize here. While we do include the namespace in the share, we don't for the erasure data. the nmt wrapper accounts for that and is therefore included in the plugin.

Reply:
@adlerjohn do you mean the recent change in the NMT lib? (ref https://github.com/celestiaorg/nmt/commit/d8fd217cf3dbbc5970c87ea474722a737d86da3c) or sth else?  @evan-forbes can you confirm if this is correct?

This is the only unresolved convo. After that I'd also lgtm.

Reply:
The extra namespace should definitely be in there. In fact, if it's not there it will break, as the nmt wrapper will use incorrect namespaces for erasure data. 
```
--- FAIL: TestDataSquareRowOrColumnRawInputParserCidEqNmtRoot (0.00s)
    --- FAIL: TestDataSquareRowOrColumnRawInputParserCidEqNmtRoot/16_leaves (0.00s)
        nmt_test.go:46: DataSquareRowOrColumnRawInputParser() unexpected error = pushed data has to be lexicographically ordered by namespace IDs: last namespace: d6b35c74f05aee9c, pushed: 19a5841380540ad8
    --- FAIL: TestDataSquareRowOrColumnRawInputParserCidEqNmtRoot/32_leaves (0.00s)
        nmt_test.go:46: DataSquareRowOrColumnRawInputParser() unexpected error = pushed data has to be lexicographically ordered by namespace IDs: last namespace: 48626b3bb56ad10a, pushed: 3e051c41caa32628
    --- FAIL: TestDataSquareRowOrColumnRawInputParserCidEqNmtRoot/extended_row (0.00s)
        nmt_test.go:46: DataSquareRowOrColumnRawInputParser() unexpected error = pushed data has to be lexicographically ordered by namespace IDs: last namespace: de80169054c0072a, pushed: 32c75158bbe36514
FAIL
exit status 1
FAIL    github.com/celestiaorg/celestia-core/pkg/da/ipfs/plugin 0.031s
```